substrate/frame/utility/src/lib.rs

```diff
@@ -504,30 +528,31 @@ pub mod pallet {
 			// Add weight of the main call
 			weight = weight.saturating_add(extract_actual_weight(&main_result, &info));
 
-			if let Err(_main_call_error) = main_result {
+			if let Err(main_error) = main_result {
 				// If the main call failed, execute the fallback call
 				let fallback_info = fallback.get_dispatch_info();
 
 				let fallback_result = if is_root {
 					fallback.dispatch_bypass_filter(origin.clone())
 				} else {
-					fallback.dispatch(origin.clone())
+					fallback.dispatch(origin)
 				};
 
 				// Add weight of the fallback call
 				weight =
 					weight.saturating_add(extract_actual_weight(&fallback_result, &fallback_info));
 
-				if let Err(_fallback_error) = fallback_result {
+				if let Err(fallback_error) = fallback_result {
 					// Both calls have faild.
+					Self::deposit_event(Event::IfElseBothFailure { main_error: main_error.error, fallback_error: fallback_error.error });
 					return Err(Error::<T>::InvalidCalls.into())
```

Comment:
We can also do the refund in case of failure.
```suggestion
					return Err(DispatchErrorWithPostInfo { error: Error::<T>::InvalidCalls.into(), post_info: Some(weight).into() })
```

Reply:
Would like to know where the refund happens